elastic/apm-server
# Add support for histograms to metrics intake

## Motivation/summary

Extend the metricset schema with additional fields on "samples": type, unit, values, and counts. For now, the only type that is honoured is "histogram", and units are not honoured at all. We need a resolution on https://github.com/elastic/elasticsearch/issues/72536 to handle gauge/counter and units.

If type is "histogram", then you are expected send "counts" and "values". Counts must be a list of non-negative integers. Values must be a list of (any) numbers; the values must be provided in ascending order.

With this change, one of "value" or "values" must be supplied. Previously "value" was required, so sending a histogram to an older server will result in a validation error. If "values" is supplied, "counts" must also be supplied.

## Checklist

~- [ ] Documentation has been updated~

## How to test these changes

Once one of our agents has implemented support for sending histogram metrics (probably via the Prometheus API?), use one to send to APM Server and ensure the metric is stored in a histogram field with expected counts/values.

## Related issues

Closes #4884
Closes #3195
